src/react.fs

```diff
@@ -6,21 +6,71 @@ open Fable.Core
 
 [<RequireQualifiedAccess>]
 module Program =
-    open Fable.Import.Browser
     module R = Fable.Helpers.React
 
+    type private SingleObservable<'T>() =
+        let mutable listener: IObserver<'T> option = None
+        member __.Trigger v =
+            listener |> Option.iter (fun lis -> lis.OnNext(v))
+        interface IObservable<'T> with
+            member __.Subscribe w =
+                listener <- Some w
+                { new IDisposable with
+                    member __.Dispose() = listener <- None }
+
+    type [<Pojo>] private Props<'S, 'Msg> =
+        { view: 'S -> ('Msg->unit) -> ReactElement
+          observable: IObservable<'S * ('Msg->unit)>
+          initState: 'S * ('Msg->unit) }
+
+    type [<Pojo>] private State<'S, 'Msg> =
+        { model: 'S
+          dispatch: 'Msg->unit }
+
+    type private ReactCom<'S, 'Msg>(p) =
+        // Don't keep properties to release memory of initial state
+        inherit Component<obj, State<'S, 'Msg>>(null)
+        let view = p.view
+        let obs = p.observable
+        let mutable disp = None
+        do base.setInitState { model = fst p.initState; dispatch = snd p.initState }
+
+        member this.componentDidMount() =
+            disp <- Some (obs.Subscribe(fun (model, dispatch) ->
+                this.setState { model = model; dispatch = dispatch }))
+
+        member __.componentWillUnmount() =
+            disp |> Option.iter (fun d -> d.Dispose())
+
+        member this.shouldComponentUpdate(_, nextState) =
+            not(obj.ReferenceEquals(this.state.model, nextState.model))
+
+        member this.render() =
+            view this.state.model this.state.dispatch
+
+    open Fable.Import.Browser
+
     /// Setup rendering of root React component inside html element identified by placeholderId
     let withReact placeholderId (program:Elmish.Program<_,_,_,_>) =
+        let mutable rootEl = None
```

Comment:
local state might break hmr, https://github.com/fable-elmish/react/pull/6/files#diff-6e208d7e773cd65ba4c7e6cf6adde006R36  maybe it's better to be put in module level?

Reply:
Makes sense, thanks! I was also thinking that a boolean flag (like `let mutable initialized = false`) is probably enough here :+1:

Reply:
I tried move the local state to global, remove `shouldComponentUpdate`, but it's strange that I cannot get HMR working with you implementation. you can try it with my demo repo by replace `paket-files/zaaack/react/src/react.fs` with yours and edit `Home/View`: https://github.com/zaaack/elmish-react-demo

Reply:
Hi @zaaack! I just did that: cloned your demo and replaced react.fs with the [code here](https://github.com/alfonsogarciacaro/ui-perf/blob/b95940ec6c654414cd2a9800d589db57c0248cfe/implementations/elmish-react-setstate/react.fs) (also uncommented the `|> Program.withHMR` line in App.fs) and HMR worked without problem. The only change needed is to remove `shouldComponentUpdate` so I guess we can wrap it `#if !DEBUG` to keep it in Release mode. Note I didn't have to make any state global.

Reply:
@alfonsogarciacaro I tried what you said before, but the view doesn't change if I edit a string child of a ReactElement, and also no errors.

Reply:
Not sure what we're doing differently because for me it's changing.

![react-setstate](https://user-images.githubusercontent.com/8275461/33122605-fee4e0ba-cf77-11e7-9496-b32e8501bd01.gif)

However, it's true HMR doesn't work completely. Updating React elements is fine for me but if I update the logic (e.g. the incremental value in Counter/State.fs) the change is not reflected in the app unless I manually refresh (this works with the current elmish/react package).

Reply:
👍  It's working now after I update to your current `react.fs`, I guess it's because the react.fs in my here is too and used a captured view:
```F#
let view = this.props.view

//...
member this.render() =
            view this.state.model this.state.dispatch
```

Now is:
```F#
member this.render() =
            this.props.view this.state.model this.state.dispatch
```

Reply:
Any clue [about this](https://github.com/fable-elmish/react/pull/10#discussion_r152523369) @MangelMaxime?

Reply:
I think because, you use a stateful components at the root. If I remember, well when I did my search about React + HMR, they where having a "debug root component" and a "production root component".

Idea being, the "bug root component" was the one supporting HMR directly and knowing how to update itself from it. Like using an existing state and forcing a whole evaluation of the react components.


Current HMR implementation is really basic. It's using a private state inside the `HMR.fs` file and re-evaluate the application from scratch like if you were doing a "F5". Because the "application state" is inside the HMR.fs module then we maintain it. Then because, we use React to make a diff between a DOM "non-empty" element and a VDOM, react detect that's the DOM element is already constructed with React so it's making a diff between this two.

I remember you (@alfonsogarciacaro) provided me a link to a github sample with React + HMR. I think it's in this repo that I saw the `index.debug.js` and `index.production.js` thing.

Reply:
Thanks a lot for the pointers @MangelMaxime! OK, [this seems to work](https://github.com/fable-elmish/react/commit/e3b56341dd1e96c861aa0d7cf57540e44505c701). What I don't understand is why _the old way_ doesn't cause the jumping cursor issue when using `Value` for an input element. Is this because of hot reloading? Or for not using `requestAnimationFrame`? Or it has been fixed in React? 🤔

Reply:
Don't know ..

Reply:
I would vote for  `not using requestAnimationFrame`...

Reply:
@alfonsogarciacaro I tried my react.fs and hmr work for update. And also yours after make local state to global state like this:

(Sorry for the downcast, I don't know how to get around `Value restriction` and just for a quick example,  so ...  😅)

```F#
open Fable.Import.Browser

    let mutable observable = None :> obj

    /// Setup rendering of root React component inside html element identified by placeholderId
    let withReact placeholderId (program:Elmish.Program<_,_,_,_>) =

        let setState model dispatch =
            match (observable :?> SingleObservable<_> option) with
            | Some obs ->
                obs.Trigger(model, dispatch)
```

Reply:
I think I'm getting a bit lost ;) I don't see any difference by placing `observable` on the top level. But yes, I can confirm `requestAnimationFrame` is what causes the jumping cursor issue, if I add it to [these lines](https://github.com/fable-elmish/react/blob/e3b56341dd1e96c861aa0d7cf57540e44505c701/src/react.fs#L61-L63) then the problem appears again.

Reply:
After I making local state to global, the new-added `printfn("update")` in update function worked, before it doesn't print anything.
![untitled3](https://user-images.githubusercontent.com/5233940/33164027-051e8cf2-d06c-11e7-85ba-46e512455b57.gif)